tslearn/nonmyopic.py

```diff
@@ -0,0 +1,509 @@
+            The cost function that will be use to achieve our classification purpose
+        Returns
+        -------
+        int: the class which is predicted
+        int : the time of the prediction
+        float : the probability used for computing the cost
+        float : the loss when classifying
+        """
+        time_stamp = xt.shape[0]
+        xt = np.reshape(xt, (1, time_stamp))
+        time_prediction = self.minimum_time_stamp
+        stop_criterion = False
+        while stop_criterion is False:
+            minimum_tau, minimum_loss = self.minimize_integer(
+                end_of_time=time_stamp - time_prediction,
+                function_to_minimize=cost_function,
+                xt=xt[:, :time_prediction],
+            )
+            if time_prediction == time_stamp:
+                result = self.classifier_[-1].predict(xt)
+                result_proba = self.classifier_[-1].predict_proba(xt)
+                stop_criterion = True
+                minimum_tau, minimum_loss = self.minimize_integer(
+                    end_of_time=time_stamp - time_prediction,
+                    function_to_minimize=cost_function,
+                    xt=xt[:, :time_prediction],
+                )
+                loss_exit = minimum_loss
+            elif minimum_tau == time_prediction:
+                result = self.classifier_[
+                    time_prediction - self.minimum_time_stamp
+                ].predict(xt[:, :time_prediction])
+                result_proba = self.classifier_[
+                    time_prediction - self.minimum_time_stamp
+                ].predict_proba(xt[:, :time_prediction])
+                loss_exit = minimum_loss
+                stop_criterion = True
+            else:
+                time_prediction = time_prediction + 1
+        return result, time_prediction, result_proba, loss_exit
+
+    @staticmethod
+    def avg_prediction(time_predictions):
+
+        avgerage_time = sum(time_predictions) / len(time_predictions)
+        variance_time = np.var(time_predictions)
+        return avgerage_time, variance_time
+
+    @staticmethod
+    def accuracy_prediction(Y_test, Y_pred_test):
+
+        good_prediction = 0
+        for class_index in range(0, len(Y_test)):
+            if Y_test[class_index] == Y_pred_test[class_index]:
+                good_prediction = good_prediction + 1
+        return good_prediction / len(Y_test)
+
+    def predict(self, X):
+        """
+        Predicts the classes of the series of the test dataset.
+
+        Parameters
+        ----------
+        X_test : Array-like
+            The test dataset
+
+        Returns
+        -------
+        Vector : the predicted classes
+        """
+
+        Y_pred = []
+
+        for t in range(0, X.shape[0]):
+            new_classif, new_time, new_proba, new_cost = self.classification(
+                X[t], self.cost_function
+            )
+            Y_pred.append(new_classif)
+        return Y_pred
```

Comment:
This one should also return the time at which prediction is made, otherwise one cannot retrieve that information.
Also, `t` is a **VERY** bad variable name for something that does not correspond to a time index but rather a series index.